### PR TITLE
Union Types in TypeScript

### LINKED ISSUE #5: Union Types

### PR DESCRIPTION
**Pull request by alexisthedev**
Union Types checklist:

- [x] Pass types to generator
- [x] Translate union type representation to actual TypeScript
- [x] Assign constant values to union types through generator (replace bottom constant use)
- [x] Create union types with previously generated types passed from generator (eg. classes)

### CODE REVIEW

**Comment by alexisthedev** [comment]
Ignore commits ecc77a3 to 8484880 they appear here from merging main into this branch.

Actual changes are represented from and after commit 3974d64

**Comment by alexisthedev** [comment]
Before this PR is ready for review, a problem where the generator object passed to the builtin factory (in order to generate the union types) can not resolve its previous non-builtin generated types (eg. classes).

The next commit will address this issue, so union types that contain non-builtin types can be generated.

**Comment by alexisthedev** [comment]
At this stage, Hephaestus can generate Union Types with previously generated types of the program (eg. classes).

Next up, on the translator-side of things, we need to correctly represent the classes with type parameters as type of the union.

For example a class `Foo<T, V extends number>` can not stand in some union as `Foo | boolean` but should be along the lines of `Foo<string, number> | boolean`. I think this can be done through accessing the program context through the translator.

Also: a somewhat rare bug has arisen where Hephaestus incorrectly assigns union types to other union types even though the appropriate type overlap does not exist. This specifically happens when union types are used to instantiate a type parameter, which is quite familiar to the bug we are currently facing on #13 Type Aliases! I will dig into this, and perhaps we can solve two problems at once.

**Comment by alexisthedev** [comment]
After adding unit tests for TypeScript and dwelling on the issue described in my previous message, the root of the problem was correctly identified and fixed.

However, support for union types with program-generated types (eg. classes) has been halted as we start planning a type-relation refactoring. When testing this form of union type, a solid mechanism of TypeScript's structural typing must first be added to Hephaestus' codebase.

This is a significant and exciting step for Hephaestus! The next commits will focus on this subject, arguably the biggest and most ambitious addition since starting to test the TypeScript compiler with Hephaestus!

**Comment by alexisthedev** [comment]
The last commit, 3fecc9a, focused on a problem with this feature I have been facing for some time now. Ended up implementing a major refactor in type unification when it comes to union types. The procedure is complex yet explained thoroughly via the comments and the function descriptions.

The next step is to handle the cases of parameterized types inside of union types, which require checking special cases in order to substitute their own type variables as well when doing type unification in a union type!

File: src/ir/builtins.py [hunk 1]
```diff
@@ -138,9 +138,19 @@ def get_function_types(self, max_parameters):
     def get_nothing(self):
         raise NotImplementedError
 
-    def get_constant_candidates(self):
+    def get_dynamic_types(self, gen_object):
```
Comment (theosotr) [inline src/ir/builtins.py]: I don't like the name `dynamic_types`. Can we think of a better name? `compound types` maybe?

File: src/ir/decorators.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+def two_way_subtyping(is_subtype):
```
Comment (theosotr) [inline src/ir/decorators.py:1]: Move this function to the `type_utils.py` file.

File: src/ir/types.py [hunk 3]
```diff
@@ -66,6 +67,19 @@ def has_type_variables(self):
     def is_subtype(self, other: Type):
         raise NotImplementedError("You have to implement 'is_subtype()'")
 
+    def dynamic_subtyping(self, other: Type):
```
Comment (theosotr) [inline src/ir/types.py]: Again, can we think of a better name?

File: src/ir/types.py [hunk 4]
```diff
@@ -107,6 +124,19 @@ def get_supertypes(self):
                     stack.append(supertype)
         return visited
 
+    def substitute_type_args(self, type_map,
+                             cond=lambda t: t.has_type_variables()):
```
Comment (theosotr) [inline src/ir/types.py]: Why do we have this method inside the generic class `Type`?. I mean a generic type has nothing to do with type arguments.

File: src/ir/typescript_types.py [hunk 5]
```diff
@@ -402,14 +424,375 @@ def gen_number_literal(self):
...
+            delete the pair with key string from our possible_subs dict.
+
+        (4) We sort the dict again, it now looks like this:
+                sorted_type_subs = [(number, {Y})]
+
+        (5) Assign the substitution Y: number and repeat the rest of step (3)
+
+        (6) The possible_subs dict is now empty, so we return our substitution dictionary
+                return {X: string, Y: number}
+        """
```
Comment (theosotr) [inline src/ir/typescript_types.py:677]: Nice docs here!

File: src/ir/typescript_types.py [hunk 6]
```diff
@@ -402,14 +424,375 @@ def gen_number_literal(self):
...
+    def __eq__(self, other):
+        return (self.__class__ == other.__class__ and
+                self.name == other.name and
+                set(self.types) == set(other.types))
+
+    def __hash__(self):
+        return hash(str(self.name) + str(self.types))
+
+
+class UnionTypeFactory:
```
Comment (theosotr) [inline src/ir/typescript_types.py]: `UnionTypeFactory` => `UnionTypeFactory(object)`.

File: src/ir/typescript_types.py [hunk 7]
```diff
@@ -402,14 +424,375 @@ def gen_number_literal(self):
...
+        """ Returns a previously created union type
+            or a newly generated at random.
+
+            If there are previously generated union types
+            and they have not exceeded the limit, we make a
+            probabilistic choice on whether to pick one of
+            the already generated types or create a new one.
+
+        """
+        return self.gen_union_type(gen_object)
```
Comment (theosotr) [inline src/ir/typescript_types.py]: Remove this `return` statement.

File: tests/test_typescript.py [hunk 8]
```diff
@@ -24,3 +29,193 @@ def test_type_alias_with_literals2():
...
+    t1 = tst.UnionType([tst.NumberType(), tst.NumberLiteralType(100), tst.BooleanType(), tst.StringLiteralType("foo"), tst.StringType()])
+
+    t_param1 = tp.TypeParameter("T", bound=tst.StringType())
+    helper_union = tst.UnionType([tst.BooleanType(), tst.StringType()])
+    t_param2 = tp.TypeParameter("G", bound=helper_union)
+    t2 = tst.UnionType([tst.NumberType(), t_param1, t_param2])
+
+    # Unify t1: number | 100 | boolean | "foo" | string
+    # with t2: number | T extends string | G extends (boolean | string)
+    # Result should be {T: StringType, G: BooleanType}
```
Comment (theosotr) [inline tests/test_typescript.py:240]: Can you add a comment (like that one) above each test case so that it is easier for us to determine what we test?

**Review by theosotr — CHANGES_REQUESTED** [review]
Good work! see comments below

File: src/ir/decorators.py (hunk 2 above)
Comment (alexisthedev) [reply src/ir/decorators.py:1]: Not possible because of impassable circular import.

File: src/ir/builtins.py (hunk 1 above)
Comment (alexisthedev) [reply src/ir/builtins.py]: Done!

File: src/ir/typescript_types.py (hunk 6 above)
Comment (alexisthedev) [reply src/ir/typescript_types.py]: Done as suggested by PEP8. The reason it is recommended is compatibility between Python 3 and 2. [Detailed explanation](https://stackoverflow.com/questions/4015417/why-do-python-classes-inherit-object).

**Pull request merged**